src/vs/workbench/contrib/welcome/page/browser/welcomePage.contribution.ts

```diff
@@ -35,13 +37,33 @@ Registry.as<IConfigurationRegistry>(ConfigurationExtensions.Configuration)
 		}
 	});
 
-Registry.as<IWorkbenchContributionsRegistry>(WorkbenchExtensions.Workbench)
-	.registerWorkbenchContribution(WelcomePageContribution, LifecyclePhase.Restored);
+class WelcomeContributions {
+	constructor(
+		@IConfigurationService configurationService: IConfigurationService,
+	) {
+		const previewFeaturesEnabled: boolean = configurationService.getValue('workbench')['enablePreviewFeatures'];
+		if (previewFeaturesEnabled) {
+			Registry.as<IWorkbenchContributionsRegistry>(WorkbenchExtensions.Workbench)
+				.registerWorkbenchContribution(WelcomePageContribution, LifecyclePhase.Restored);
+
+			Registry.as<IWorkbenchActionRegistry>(ActionExtensions.WorkbenchActions)
+				.registerWorkbenchAction(SyncActionDescriptor.create(WelcomePageAction, WelcomePageAction.ID, WelcomePageAction.LABEL), 'Help: Welcome', localize('help', "Help"));
+
+			Registry.as<IEditorInputFactoryRegistry>(EditorExtensions.EditorInputFactories).registerEditorInputFactory(WelcomeInputFactory.ID, WelcomeInputFactory);
+		} else {
+			Registry.as<IWorkbenchContributionsRegistry>(WorkbenchExtensions.Workbench)
+				.registerWorkbenchContribution(WelcomePageContribution2, LifecyclePhase.Restored);
 
-Registry.as<IWorkbenchActionRegistry>(ActionExtensions.WorkbenchActions)
-	.registerWorkbenchAction(SyncActionDescriptor.create(WelcomePageAction, WelcomePageAction.ID, WelcomePageAction.LABEL), 'Help: Welcome', localize('help', "Help"));
+			Registry.as<IWorkbenchActionRegistry>(ActionExtensions.WorkbenchActions)
```

Comment:
Shouldn't the welcome action be the one deciding which page to use based on whether preview features are enabled? With this if a user changes whether preview features are enabled after starting ADS the `Help -> Welcome` action will still use whatever the page appropriate for the original startup value of preview features being enabled.

Reply:
This implementation will require a restart for the flag to take effect.  This seems like the lowest risk implementation for the April release at this time.